### PR TITLE
Add distributed primitives

### PR DESCRIPTION
**Pull request by gfabbretti8**
## Description

Added support for three distributed primitives (start/1,2, node/0, nodes/0) and updated the spawn so that remote spawns (spawn/4) are also possible.

## Changes

- Support for distributed primitives
- Refactor of the 'cauder_semantics_forwards.erl' file

## Checklist

- [X] Follows styleguide
- [X] Branch is up to date
- [x] Passed unit tests

## References

The formal semantics can be found here: https://gitlab.inria.fr/gfabbret/distributed-cauder-complete-semantics.git

### CODE REVIEW

**Review by SackCastellon — COMMENTED** [review]
Thanks for the PR. I have some comments:

- Because of the change to the structure of the log, the existing [case studies](https://github.com/mistupv/cauder/tree/dev/case-studies) no longer work if loading a log (trace).

- Specifying a node name should be optional and not a requirement to start a session, and the behavior of the added functions should change according to whether the node is alive or not. Specifically, if the node is not alive:
  - `erlang:node/0` should return `nonode@nohost`.
  - `slave:start/{1,2}` should crash with the reason `not_alive`.
  - `erlang:spawn/4` should return a useless PID.

- Could you also add support for the function `erlang:spawn/2`?

**Comment by gfabbretti8** [comment]
Thanks for your feedback. About the comments:

- We could retrace the case studies with the new tracer so that the final trace will also have the atom success/fail in the spawn log
- It is possible to make the node parameter optional, how should we treat the exception 'no_alive'? Since exceptions are still not handled prompting a message informing the user that an exception occurred and interrupting the debugging session could be fine?
- Yes, I will do it as soon as possible

**Comment by german-vidal** [comment]
Hi Giovanni,

> 	• We could retrace the case studies with the new tracer so that the final trace will also have the atom success/fail in the spawn log

Is this extended tracer available? And, if so, have you updated it in github?

The new extension looks great, but ideally the extensions should be 
backwards compatible, at least until the new tracer is available
(and even then I'd still prefer it to be backwards compatible).

BTW, we're also working on a new tracer. Let us know what are the
required new events and a brief description, so that we can also
take them into account in our implementation.

Perhaps we can move this topic to the cauder mailing list...

**Comment by SackCastellon** [comment]
> * It is possible to make the node parameter optional, how should we treat the exception 'no_alive'? Since exceptions are still not handled prompting a message informing the user that an exception occurred and interrupting the debugging session could be fine?

Yes, showing a message and interrupting the session should be enough for now.

**Comment by gfabbretti8** [comment]
Hi German,

I updated the tracer on the official repo, under the 'with_dbg' branch. 

I also performed the other changes that Juan José asked me (i.e., making the node parameter optional and support the spawn/2), while about the log of the spawn I am not so sure on how to proceed.

The biggest problem here is the that the structure of the log is fixed so having backwards compatibility translates to supporting one tuple of the form `{spawn, pid}` and one of the form `{spawn, {node, succ/fail, pid}}`, so every function that manipulate such log must be able to deal with both of its forms. 

Moreover, with the growth of CauDEr more and more constructs will be added and I fear that we could face a similar situation again in the future. Perhaps it may be a good idea to consider a more flexible data structure for the log (e.g., a map) so that the lack of an information can be treated as the information itself. For instance, in this case the lack of the value `succ/fail` can safely be considered as a `succ`.

I believe that such approach should work also in the future because given two semantics of CauDEr S and S', such that S' extends S, then we should always be able to see S as a simpler case of a S'.

Please let me know what you think and what is the best way to proceed in your opinion.

PS. I will send an email on the devs' mailing list about what information is required from the tracer for the distributed semantics.

**Comment by german-vidal** [comment]
Hi,

> I updated the tracer on the official repo, under the 'with_dbg' branch.

Great, thanks! We have checked it, and it seems to work fine with the old examples. 

BTW, I assume that "receive" events now correspond to a receive statement 
consuming a message (as in the old tracer) rather than the delivery of the 
message to the process' mailbox. Right?

> I also performed the other changes that Juan José asked me (i.e., making the node parameter optional and support the spawn/2), while about the log of the spawn I am not so sure on how to proceed.
> 
> The biggest problem here is the that the structure of the log is fixed so having backwards compatibility translates to supporting one tuple of the form {spawn, pid} and one of the form {spawn, {node, succ/fail, pid}}, so every function that manipulate such log must be able to deal with both of its forms.
> 
> Moreover, with the growth of CauDEr more and more constructs will be added and I fear that we could face a similar situation again in the future. Perhaps it may be a good idea to consider a more flexible data structure for the log (e.g., a map) so that the lack of an information can be treated as the information itself. For instance, in this case the lack of the value succ/fail can safely be considered as a succ.
> 
> I believe that such approach should work also in the future because given two semantics of CauDEr S and S', such that S' extends S, then we should always be able to see S as a simpler case of a S'.

Yes, I agree that we're going to need a better definition for traces so that
new versions of the log with more features and/or more info do not require 
modifying the current version of CauDEr necessarily (or, at least, the changes 
are well located in a single place). 

Actually, we are also working on a new tracer including line numbers, four events
(spawn, send, deliver and receive), and will eventually consider other extensions
like shared-memory, input/output, process' events, etc.

To summarize, yes, let's go ahead with the update. Juanjo, please add some info
in the CauDEr site pointing out to the right version of the tracer. Perhaps we
could integrate the tracer into the CauDEr github? I'm a bit worried that the
connection between CauDEr and the tracer is not very explicit and a user might
get confused now if he/she downloads only CauDEr (or CauDEr and an incompatible
version of the tracer). What do you think?

Cheers,
German

PS. BTW, Giovanni, I'm not sure I follow your arguments about the use of maps.
Could you elaborate it? In principle, we consume logs sequentially, so the 
advantages of using dicts/maps/etc are not obvious to me. What had you in mind?

**Comment by gfabbretti8** [comment]
Hi, 

> BTW, I assume that "receive" events now correspond to a receive statement
consuming a message (as in the old tracer) rather than the delivery of the
message to the process' mailbox. Right?

Yes, exactly.


> PS. BTW, Giovanni, I'm not sure I follow your arguments about the use of maps.
> Could you elaborate it? In principle, we consume logs sequentially, so the
> advantages of using dicts/maps/etc are not obvious to me. What had you in mind? 

My idea was to still have a sequence of log entries but where each log is a 'more dynamic structure' than a tuple.
My intuition is that if we move from a semantics S to a more complicated one S' then it should always exists an automatic translation of the information in the log.
For instance in this example we have the following automatic translation:
`{spawn, pid}  --> {spawn, {nonode@nohost, succ, pid}}'

If the log entry is represented by a map then when a value is not defined we can just give it the appropriate default value.

But perhaps an even better idea would be to do this translation once and for all when the trace gets loaded, in this case we will not have to worry about the user using the "right" tracer (which is a legit concern). 
That could be also the single places where to locate all the changes that you mentioned.

In conclusion, if you think that such approach is a good idea I will update this version of CauDEr to update eventual log entries with the old form to the new one, so there will not be the necessity to explain which one is the "right" version of the tracer.
Let me know what you think.

Best,
Giovanni.

**Comment by SackCastellon** [comment]
> Juanjo, please add some info in the CauDEr site pointing out to the right version of the tracer. Perhaps we could integrate the tracer into the CauDEr github? I'm a bit worried that the connection between CauDEr and the tracer is not very explicit and a user might get confused now if he/she downloads only CauDEr (or CauDEr and an incompatible version of the tracer). What do you think?

Ok, I'll add a section to the README about creating logs using the tracer.
About integrating the tracer into this repo, it seem appropriate given that the tracer is only used to create logs for CauDEr. It could also have its own UI, maybe integrated into CauDEr?
I can create a PR later.

**Review by SackCastellon — APPROVED** [review]
